src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,177 @@
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findByIndex(index);
+        T result = node.item;
+        unlink(node);
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        if (findByItem(object) != null) {
+            Node<T> node = findByItem(object);
+            unlink(node);
+            size--;
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Cann`t find index " + index);
+        }
+    }
+
+    private void checkToAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Cann`t find index " + index);
+        }
+    }
+
+    private Node<T> findByIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode;
+        if (size / 2 > index) {
+            currentNode = first;
+            for (int i = 1; i <= index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = last;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private Node<T> findByItem(T value) {
+        Node<T> current = first;
+        while (current != null) {
+            if (current.item == value || (current.item != null && (current.item).equals(value))) {
+                return current;
+            }
+            current = current.next;
+        }
+        return null;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev != null && node.next != null) {
+            (node.next).prev = node.prev;
+            (node.prev).next = node.next;
+        }
+        if (node.next == null) {
+            last = node.prev;
+            if (node.prev != null) {
+                (node.prev).next = null;
+            }
+        }
+        if (node.prev == null) {
+            first = node.next;
+            if (node.next != null) {
+                (node.next).prev = null;
+            }
+        }
+    }
+
+    private void linkWithLast(T value) {
+        Node<T> node = new Node<>(last, value, null);
+        if (last == null) {
+            first = node;
+        } else {
+            last.next = node;
+            node.prev = last;
+        }
+        last = node;
+    }
+
+    private void linkNode(T value, Node<T> nextNode) {
+        Node<T> node = new Node<>(null, value, nextNode);
+        if (nextNode.prev == null) {
+            first = node;
+        } else {
+            (nextNode.prev).next = node;
```

Comment:
The parentless is redundant
```suggestion
            nextNode.prev.next = node;
```

Reply:
agree